src/lava/utils/float2fixed.py

```diff
@@ -1,26 +1,719 @@
+                # Create copy of conv_data.
+                conv_data_copy = conv_data.copy()
+
+                # Get scale domain and remove it from dictionary.
+                scale_domain = conv_data_copy.pop('scale_domain')
+                # Remove other information not needed anymore.
+                conv_data_copy.pop('constant')
+                conv_data_copy.pop('exp_var')
+
+                if conv_data_copy['num_bits_exp'] is None:
+                    conv_data_copy['num_bits_exp'] = 0
+                scale_domains[scale_domain][p_id][var] = conv_data_copy
+
+        return scale_domains
+
+    def _find_scaling_factors(self) -> dict:
+        """Determine the optimal scaling function for the float- to fixed-point
+        mapping. The function must be consistent between Variables of one scale
+        domain of be chosen such that for every Variable the numerical
+        fixed-point representation can be represented with the given
+        constraints. The optimal scaling function will be stored in a
+        dictionary with keys scale domains.
+
+        Returns
+        -------
+        scaling_funct_dict: dict
+            Dictionary containing the optimal scaling function for each scale
+            domain.
+        """
+        # Initialize scaling function dictionary.
+        scaling_factors = {}
+        slopes = []
+
+        for scale_domain, scale_data_domain in self.scale_domains.items():
+            scaling_factors[scale_domain] = {}
+            for p_id, scale_data_p_id in scale_data_domain.items():
+                for var, scale_data in scale_data_p_id.items():
+
+                    bits = scale_data['precision_bits']
+                    impl_shift = scale_data['implicit_shift']
+                    num_bits_exp = scale_data['num_bits_exp']
+
+                    # Get range representable given bit constraints and
+                    # signedness.
+                    if scale_data['signedness'] == 's':
+                        rep_range = np.array([-1 * 2 ** (bits - 1),
+                                              2 ** (bits - 1) - 1])
+                    elif scale_data['signedness'] == 'u':
+                        rep_range = np.array([0, 2 ** bits - 1])
+
+                    # Apply implicit shift.
+                    rep_range = np.left_shift(rep_range, impl_shift)
+
+                    # Apply maximal shift due to exponent.
+                    rep_range = np.left_shift(rep_range, 2 ** num_bits_exp - 1)
+
+                    # Get range of Variables in floating point domain.
+                    domain = scale_data['domain']
+
+                    # If domain of Variable has been recorded apply quantile.
+                    if var in self.monitors[p_id].keys():
+                        lower_val = np.quantile(domain, self.quantiles[0])
+                        higher_val = np.quantile(domain, self.quantiles[1])
+                        dyn_range = np.array([lower_val, higher_val])
```

Comment:
Also here, how to represent ranges? Array, list, tuple? I would argue for tuples.

Reply:
True, addressed in feb0b3aab0d2bc85d74e6380035de06d84b4f615.